### PR TITLE
Manage infrastructure using terraform

### PR DESCRIPTION
**Pull request by mogul**
Ref #582 #583 #584 

This PR introduces the `terraform/` subdirectory, which is intended to be used for managing all "non-code" configuration for our environments: Services, routes, access, security groups, etc. It generally implements the ADR described in #660.

- A `bootstrap` directory was used for setting up the remote Terraform state in an S3 service in the `production` space, as described in the README.md.
- There's a directory with a module for each managed environment/space (`dev`, `staging`, `production`, and `management` initially) per [this guidance from Truss](https://truss.works/blog/building-for-the-future-with-terraform).
  - The `management` module ~~provisions and~~ manages access to the spaces
  - The other modules are intended to manage non-code assets in each space; right now there's nothing managed by Terraform, but a commented-out DB service shows what that should look like.
- There are GitHub workflows for previewing changes by posting the `terraform plan` on PRs
- A reusable `terraform apply` workflow is here, but it's not actually called by anything in `test_and_deploy.yml` yet; I'm waiting for that to stabilize before I submit another PR to hook it in!

### CODE REVIEW

**Comment by geekygirlsarah** [comment]
Ahhh Terraform, we meet again...

**Comment by tadhg-ohiggins** [comment]
@mogul and I discussed some things we're looking out for, and came up with this:

----
Competing interests:

1.  We want different Terraform directories so that we can support e.g different database plans for `dev` versus `staging` or `production` (more storage/compute resources/persistence/etc.).
2.  We want to avoid a situation where a code change that depends on an infrastructure change goes to `dev` and look great, but only has the Terraform infrastructure rules for `dev` in it. Once pushed to `staging`, this would break because it doesn’t contain the corresponding Terraform rules for `staging` but nothing in our process would warn us of this prior. (This also applies for things looking fine in `staging` but where we forgot the rules for `production`.)

We can address both of these interest via the following Terraform setup:

*   Use Terraform modules extensively in config.
*   The environments only specify where they are different from the defaults.
*   The defaults are all in a shared module.

This is an example of how that might look:

```terraform
module "fac-dev" {
  source = "../modules/fac"
  database_plan = "micro-psql"
}
```

The above example uses the default from `../modules/fac` but overrides the value of `database_plan` to use a smaller instance for the `dev` environment.

If we want something to be different between environments, we have to explicitly allow for customization by providing an input variable—Terraform modules must specify which of their variables can be overridden.

Example of modules/fac (where we concentrate all the config details with staging/prod defaults, but allow for override through variables):

```terraform
variable "database_plan" {
  type        = string
  description = "name of the service plan name to create"
  default     = "xlarge-gp-psql-redundant"
}

module "database" {
  source = "../modules/database"

  …

  rds_plan_name    = var.database_plan
}
```

**Comment by mogul** [comment]
Now rebased on #639 since we want terraform changes for a given environment to be deployed right before code changes, and @LindsayYoung's PR isn't merged yet. Will mark non-draft once 639 is in.

**Comment by mogul** [comment]
FYI: I rebased this onto `main` just now so we can potentially land this around the other changes in flight (to refactor `test_and_deploy.yml` to use environments).

File: terraform/README.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  |- variables.tf
+  |- run.sh
+  |- teardown_creds.sh
+  |- import.sh
+- <env>/
+  |- main.tf
+  |- providers.tf
+  |- secrets.auto.tfvars
+  |- variables.tf
+- shared/
```
[the remark this answers is not in the document]
Comment (mogul) [reply terraform/README.md:119]: I'll be making big changes to the bootstrap process in another PR that will revamp the README entirely so I'm going to hold off on here for now.

File: terraform/shared/modules/base/variables.tf [hunk 2]
```diff
@@ -0,0 +1,26 @@
+variable "cf_org_name" {
```
Comment (mogul) [inline terraform/shared/modules/base/variables.tf:1]: Add a note here indicating that defaults are production defaults, and they're exposed as variables to enable environments to specify alternate config when needed. Reference the production module as the "no config necessary" invocation.

File: terraform/management/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Management environment
+
+The "management" is a "meta" configuration, in that it specifies
+configuration pertaining to spaces and users across environments. In an
+initial bring-up/contingency recovery situation, this is the first environment
+to apply.
+
+Among other things, this module...
+- configures up the spaces for dev, staging, and production
```
Comment (mogul) [inline terraform/management/README.md:9]: Not yet, due to provider bug; to be addressed in bootstrap PR, so ~~strikethrough~~ here until that's fixed.

File: terraform/management/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+- configures user access to those spaces
+- configures application security groups (ASGs) for the spaces
```
Comment (mogul) [inline terraform/management/README.md:11]: Again, ~~strikethrough~~ due to the upstream provider bug

File: terraform/management/config.tf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    "sarah.withee@gsa.gov",
+    "tadhg.ohiggins@gsa.gov",
+    "timothy.ballard@gsa.gov",
+  ]
+
+  managers = [
+    # The OrgManager user explicitly sets itself as a SpaceManager for each space
+    var.cf_user,
+
+    # Mirrors @GSA-TTS/FAC-admins
```
Comment (mogul) [inline terraform/management/config.tf:25]: Copy this comment up for the developers list too.

File: terraform/management/main.tf [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+# Ensure spaces exist
+resource "cloudfoundry_space" "space" {
+  for_each = toset(local.spaces)
+  name     = each.key
+  org      = data.cloudfoundry_org.org.id
+
+  # Disallow SSH access in production
+  allow_ssh = each.key != "production"
+
+  # All spaces have full public egress (for now)
```
Comment (mogul) [inline terraform/management/main.tf:20]: This comment should be on the config, not the resource

File: terraform/management/main.tf [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  asgs = [for d in data.cloudfoundry_asg.asgs : d.id]
+}
+
+# Everyone has SpaceDeveloper permission in all spaces
```
Comment (mogul) [inline terraform/management/main.tf:24]: Move comment up to the configuration

**Review by mogul — COMMENTED** [review]
Comments from my review with Tadhg just now.

**Review by tadhg-ohiggins — APPROVED** [review]
Went over this in pairing, looks good.

**Pull request merged**